wiring/inc/spark_wiring_diagnostics.h

```diff
@@ -57,120 +59,422 @@ class AbstractIntegerDiagnosticData: public AbstractDiagnosticData {
+        const auto lock = ConcurrencyT::lock();
+        val = (IntType)val_;
+        ConcurrencyT::unlock(lock);
+        return SYSTEM_ERROR_NONE;
+    }
 };
 
 template<typename EnumT>
-class EnumDiagnosticData<EnumT, AtomicLockingPolicy>: public AbstractIntegerDiagnosticData {
+class EnumDiagnosticData<EnumT, AtomicConcurrency>: public AbstractIntegerDiagnosticData {
 public:
     typedef EnumT EnumType;
 
-    EnumDiagnosticData(uint16_t id, EnumT val);
-    EnumDiagnosticData(uint16_t id, const char* name, EnumT val);
+    EnumDiagnosticData(DiagnosticDataId id, EnumT val) :
+            EnumDiagnosticData(id, nullptr, val) {
+    }
+
+    EnumDiagnosticData(DiagnosticDataId id, const char* name, EnumT val) :
+            AbstractIntegerDiagnosticData(id, name),
+            val_((IntType)val) {
+    }
+
+    EnumDiagnosticData& operator=(EnumT val) {
+        val_.store((IntType)val, std::memory_order_relaxed);
+        return *this;
+    }
 
-    EnumDiagnosticData<EnumT, AtomicLockingPolicy>& operator=(EnumT val);
-    operator EnumT() const;
+    operator EnumT() const {
+        return (EnumT)val_.load(std::memory_order_relaxed);
+    }
 
 private:
     std::atomic<IntType> val_;
 
-    virtual int get(IntType& val) override; // AbstractIntegerDiagnosticData
+    virtual int get(IntType& val) override { // AbstractIntegerDiagnosticData
+        val = val_.load(std::memory_order_relaxed);
+        return SYSTEM_ERROR_NONE;
+    }
+};
+
+template<typename EnumT, typename StorageT, typename ConcurrencyT = NoConcurrency>
+class PersistentEnumDiagnosticData:
+        public AbstractIntegerDiagnosticData,
+        private ConcurrencyT {
+public:
+    typedef EnumT EnumType;
+    typedef typename StorageT::ValueType ValueType;
+
+    PersistentEnumDiagnosticData(StorageT* storage, DiagnosticDataId id, EnumT val) :
+            PersistentEnumDiagnosticData(storage, id, nullptr, val) {
+    }
+
+    PersistentEnumDiagnosticData(StorageT* storage, DiagnosticDataId id, const char* name, EnumT val) :
+            AbstractIntegerDiagnosticData(id, name),
+            storage_(storage) {
+        storage_->init((ValueType)val);
+    }
+
+    PersistentEnumDiagnosticData& operator=(EnumT val) {
+        const auto lock = ConcurrencyT::lock();
+        storage_->value() = (ValueType)val;
+        storage_->update();
+        ConcurrencyT::unlock(lock);
+        return *this;
+    }
+
+    operator EnumT() const {
+        const auto lock = ConcurrencyT::lock();
+        const EnumT v = (EnumT)storage_->value();
+        ConcurrencyT::unlock(lock);
+        return v;
+    }
+
+private:
+    StorageT* storage_;
+
+    virtual int get(IntType& val) override { // AbstractIntegerDiagnosticData
+        const auto lock = ConcurrencyT::lock();
+        val = (IntType)storage_->value();
+        ConcurrencyT::unlock(lock);
+        return SYSTEM_ERROR_NONE;
+    }
+};
+
+template<typename ValueT, DiagnosticDataId id>
```

Comment:
I wonder how much flash space is taken each time this template is instantiated with a different ID.  Would making the ID member field be worth it?  This could totally be premature optimization that we can do later.

Reply:
yeah, I had the same concern but decided in favour of the lower backup RAM usage. At second thought, the overhead of storing an ID in the backup RAM is pretty minimal and as a bonus we're getting a slightly more reliable integrity check

Reply:
I don't mean that the ID would be stored in backup ram, but that it is stored in SRAM, so we don't need a new template instantiation for each ID.